### PR TITLE
rpc: add private & unconditional to /dial_peer

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

Allow dialing of private and unconditional peers through the RPC

Closes: #1705

### CODE REVIEW

File: rpc/client/local/local.go [hunk 1]
```diff
@@ -132,8 +132,8 @@ func (c *Local) DialSeeds(seeds []string) (*ctypes.ResultDialSeeds, error) {
 	return core.UnsafeDialSeeds(c.ctx, seeds)
 }
 
-func (c *Local) DialPeers(peers []string, persistent bool) (*ctypes.ResultDialPeers, error) {
-	return core.UnsafeDialPeers(c.ctx, peers, persistent)
+func (c *Local) DialPeers(peers []string, persistent, unconditionatl, private bool) (*ctypes.ResultDialPeers, error) {
```
Comment (melekes) [inline rpc/client/local/local.go:135]:
```suggestion
func (c *Local) DialPeers(peers []string, persistent, unconditional, private bool) (*ctypes.ResultDialPeers, error) {
```

File: rpc/client/local/local.go [hunk 2]
```diff
@@ -132,8 +132,8 @@ func (c *Local) DialSeeds(seeds []string) (*ctypes.ResultDialSeeds, error) {
...
+	return core.UnsafeDialPeers(c.ctx, peers, persistent, unconditionatl, private)
```
Comment (melekes) [inline rpc/client/local/local.go:136]:
```suggestion
	return core.UnsafeDialPeers(c.ctx, peers, persistent, unconditional, private)
```

File: rpc/client/mock/client.go [hunk 3]
```diff
@@ -138,8 +138,8 @@ func (c Client) DialSeeds(seeds []string) (*ctypes.ResultDialSeeds, error) {
 	return core.UnsafeDialSeeds(&rpctypes.Context{}, seeds)
 }
 
-func (c Client) DialPeers(peers []string, persistent bool) (*ctypes.ResultDialPeers, error) {
-	return core.UnsafeDialPeers(&rpctypes.Context{}, peers, persistent)
+func (c Client) DialPeers(peers []string, persistent, uncoditional, private bool) (*ctypes.ResultDialPeers, error) {
```
Comment (melekes) [inline rpc/client/mock/client.go:141]:
```suggestion
func (c Client) DialPeers(peers []string, persistent, unconditional, private bool) (*ctypes.ResultDialPeers, error) {
```

File: rpc/client/mock/client.go [hunk 4]
```diff
@@ -138,8 +138,8 @@ func (c Client) DialSeeds(seeds []string) (*ctypes.ResultDialSeeds, error) {
...
+	return core.UnsafeDialPeers(&rpctypes.Context{}, peers, persistent, uncoditional, private)
```
Comment (melekes) [inline rpc/client/mock/client.go:142]:
```suggestion
	return core.UnsafeDialPeers(&rpctypes.Context{}, peers, persistent, unconditional, private)
```

File: rpc/core/net.go [hunk 5]
```diff
@@ -51,19 +52,42 @@ func UnsafeDialSeeds(ctx *rpctypes.Context, seeds []string) (*ctypes.ResultDialS
 
 // UnsafeDialPeers dials the given peers (comma-separated id@IP:PORT),
 // optionally making them persistent.
-func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent bool) (*ctypes.ResultDialPeers, error) {
+func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent, uncoditional, private bool) (
```
Comment (melekes) [inline rpc/core/net.go:55]:
```suggestion
func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent, unconditional, private bool) (
```

File: rpc/core/net.go [hunk 6]
```diff
@@ -51,19 +52,42 @@ func UnsafeDialSeeds(ctx *rpctypes.Context, seeds []string) (*ctypes.ResultDialS
...
 	}
-	env.Logger.Info("DialPeers", "peers", peers, "persistent", persistent)
+
+	Ids, err := getIDs(peers)
+	if err != nil {
+		return &ctypes.ResultDialPeers{}, err
+	}
+
+	env.Logger.Info("DialPeers", "peers", peers, "persistent",
+		persistent, "uncoditional", uncoditional, "private", private)
```
Comment (melekes) [inline rpc/core/net.go:67]:
```suggestion
		persistent, "unconditional", unconditional, "private", private)
```

File: rpc/core/net.go [hunk 7]
```diff
@@ -51,19 +52,42 @@ func UnsafeDialSeeds(ctx *rpctypes.Context, seeds []string) (*ctypes.ResultDialS
...
 		}
 	}
+
+	if private {
+		if err := env.P2PPeers.AddPrivatePeerIDs(Ids); err != nil {
+			return &ctypes.ResultDialPeers{}, err
+		}
+	}
+
+	if uncoditional {
```
Comment (melekes) [inline rpc/core/net.go:81]:
```suggestion
	if unconditional {
```

File: rpc/core/net.go [hunk 8]
```diff
@@ -72,3 +96,18 @@ func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent bool) (*c
 func Genesis(ctx *rpctypes.Context) (*ctypes.ResultGenesis, error) {
 	return &ctypes.ResultGenesis{Genesis: env.GenDoc}, nil
 }
+
+func getIDs(peers []string) ([]string, error) {
+	Ids := make([]string, 0, len(peers))
```
Comment (melekes) [inline rpc/core/net.go:101]:
```suggestion
	ids := make([]string, 0, len(peers))
```

File: rpc/core/net.go [hunk 9]
```diff
@@ -72,3 +96,18 @@ func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent bool) (*c
...
+
+	for _, peer := range peers {
+
+		spl := strings.Split(peer, "@")
+		if len(spl) != 2 {
+			return nil, p2p.ErrNetAddressNoID{Addr: peer}
+		}
+		Ids = append(Ids, spl[0])
```
Comment (melekes) [inline rpc/core/net.go:109]:
```suggestion
		ids = append(ids, spl[0])
```

File: rpc/core/net.go [hunk 10]
```diff
@@ -72,3 +96,18 @@ func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent bool) (*c
...
+
+	}
+	return Ids, nil
```
Comment (melekes) [inline rpc/core/net.go:112]:
```suggestion
	return ids, nil
```

File: rpc/core/net_test.go [hunk 11]
```diff
@@ -61,16 +61,17 @@ func TestUnsafeDialPeers(t *testing.T) {
 	env.P2PPeers = sw
 
 	testCases := []struct {
-		peers []string
-		isErr bool
+		peers                                 []string
+		persistence, uncondidtiional, private bool
```
Comment (melekes) [inline rpc/core/net_test.go:65]:
```suggestion
		persistence, unconditional, private bool
```

File: rpc/core/net_test.go [hunk 12]
```diff
@@ -61,16 +61,17 @@ func TestUnsafeDialPeers(t *testing.T) {
...
-		{[]string{"d51fb70907db1c6c2d5237e78379b25cf1a37ab4@127.0.0.1:41198"}, false},
-		{[]string{"127.0.0.1:41198"}, true},
+		{[]string{}, false, false, false, true},
+		{[]string{"d51fb70907db1c6c2d5237e78379b25cf1a37ab4@127.0.0.1:41198"}, true, true, true, false},
+		{[]string{"127.0.0.1:41198"}, true, true, false, true},
 	}
 
 	for _, tc := range testCases {
-		res, err := UnsafeDialPeers(&rpctypes.Context{}, tc.peers, false)
+		res, err := UnsafeDialPeers(&rpctypes.Context{}, tc.peers, tc.persistence, tc.uncondidtiional, tc.private)
```
Comment (melekes) [inline rpc/core/net_test.go:74]:
```suggestion
		res, err := UnsafeDialPeers(&rpctypes.Context{}, tc.peers, tc.persistence, tc.unconditional, tc.private)
```

**Review by melekes — CHANGES_REQUESTED** [review]

File: rpc/core/net.go [hunk 13]
```diff
@@ -51,19 +52,42 @@ func UnsafeDialSeeds(ctx *rpctypes.Context, seeds []string) (*ctypes.ResultDialS
...
 // optionally making them persistent.
-func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent bool) (*ctypes.ResultDialPeers, error) {
+func UnsafeDialPeers(ctx *rpctypes.Context, peers []string, persistent, unconditional, private bool) (
+	*ctypes.ResultDialPeers, error) {
 	if len(peers) == 0 {
 		return &ctypes.ResultDialPeers{}, errors.New("no peers provided")
 	}
-	env.Logger.Info("DialPeers", "peers", peers, "persistent", persistent)
+
+	Ids, err := getIDs(peers)
```
Comment (melekes) [inline rpc/core/net.go:61]:
```suggestion
	ids, err := getIDs(peers)
```

File: rpc/core/net.go [hunk 14]
```diff
@@ -51,19 +52,42 @@ func UnsafeDialSeeds(ctx *rpctypes.Context, seeds []string) (*ctypes.ResultDialS
...
+		persistent, "unconditional", unconditional, "private", private)
+
 	if persistent {
 		if err := env.P2PPeers.AddPersistentPeers(peers); err != nil {
 			return &ctypes.ResultDialPeers{}, err
 		}
 	}
+
+	if private {
+		if err := env.P2PPeers.AddPrivatePeerIDs(Ids); err != nil {
```
Comment (melekes) [inline rpc/core/net.go:76]:
```suggestion
		if err := env.P2PPeers.AddPrivatePeerIDs(ids); err != nil {
```

File: rpc/core/net.go [hunk 15]
```diff
@@ -51,19 +52,42 @@ func UnsafeDialSeeds(ctx *rpctypes.Context, seeds []string) (*ctypes.ResultDialS
...
+			return &ctypes.ResultDialPeers{}, err
+		}
+	}
+
+	if unconditional {
+		if err := env.P2PPeers.AddUnconditionalPeerIDs(Ids); err != nil {
```
Comment (melekes) [inline rpc/core/net.go:82]:
```suggestion
		if err := env.P2PPeers.AddUnconditionalPeerIDs(ids); err != nil {
```

**Comment by marbar3778** [comment]
eek there is a data race detected. Will look into if its something introduced here or its being exposed via this PR

**Review by melekes — APPROVED** [review]
👍

**Pull request merged**